packages/dds/tree/src/changeset/utils.ts

```diff
@@ -0,0 +1,146 @@
+    }
+    for (let i = 0; i < lhs.length; ++i) {
+        const lhsEffect = lhs[i];
+        const rhsEffect = rhs[i];
+        if (
+            lhsEffect.id !== rhsEffect.id
+            || lhsEffect.type !== rhsEffect.type
+            || lhsEffect.excludePriorInsertions !== rhsEffect.excludePriorInsertions
+            || lhsEffect.includePosteriorInsertions !== rhsEffect.includePosteriorInsertions
+        ) {
+            return false;
+        }
+    }
+    return true;
+}
+
+export function getMarkLength(mark: T.Mark): number {
+    if (typeof mark === "number") {
+        return mark;
+    }
+    if (isAttachGroup(mark)) {
+        return mark.reduce((prev, attach) => prev + getAttachLength(attach), 0);
+    }
+    return "count" in mark ? mark.count : 1;
+}
+
+export function splitMark<TMark extends T.Mark>(mark: TMark, length: number): [TMark, TMark] {
+    const markLength = getMarkLength(mark);
+    const remainder = markLength - length;
+    if (length < 1 || markLength <= length) {
+        fail(`Unable to split mark of length ${markLength} into marks of lengths ${length} and ${remainder}`);
+    }
+    if (typeof mark === "number") {
+        return [length, remainder] as [TMark, TMark];
+    }
+    if (isAttachGroup(mark)) {
+        return splitAttachGroup(mark, length) as [TMark, TMark];
+    }
+    // The linter seems to think this cast is not needed, which seems correct, but the compiled disagrees.
+    const markObj = mark as T.ObjectMark;
+    if ("count" in mark) {
+        return [{ ...markObj, count: length }, { ...markObj, count: remainder }] as [TMark, TMark];
+    }
+    fail("Unable to split mark");
+}
+
+function splitAttachGroup<TAttach extends T.Attach>(mark: TAttach[], length: number): [TAttach[], TAttach[]] {
+    const groupA: TAttach[] = [];
+    const groupB: TAttach[] = [...mark];
+    let left = length;
+    while (left > 0) {
+        const attach = groupB.shift();
+        if (attach === undefined) {
+            fail("Discrepancy between getMarkLength and splitMark");
+        }
+        const len = getAttachLength(attach);
+        if (len <= left) {
+            groupA.push(attach);
+            left -= len;
+        } else {
+            const pair = splitAttachMark(attach, left);
+            groupA.push(pair[0]);
+            groupB.unshift(pair[1]);
+            left -= left;
+        }
+    }
+    return [groupA, groupB];
+}
+
+function splitAttachMark<TAttach extends T.Attach>(attach: TAttach, length: number): [TAttach, TAttach] {
+    const markLength = getAttachLength(attach);
+    const remainder = markLength - length;
+    if (length < 1 || markLength <= length) {
+        fail(`Unable to split mark of length ${markLength} into marks of lengths ${length} and ${remainder}`);
+    }
+    const type = attach.type;
+    switch (type) {
+        case "Bounce":
+        case "Intake":
+        case "MInsert":
+        case "MMoveIn":
+            fail("Unable to split mark");
```

Comment:
Might be cleaner to just make this the default case.

Reply:
It would make the code slightly simpler, but it would also make it easier to break since someone adding a mark type would no longer get a compile error. I'd rather they get the compile error.